# Introduction

Nous disposons de deux ensembles de données. Un spécifique à la Berce du Caucase et un autre contenant de nombreuses informations sur plusieurs néophytes.

Dans ce notebook, nous effectuerons une Analyse Exploratoire de ces  fichiers, afin de déterminer si l'un d'entre eux est plus prometteur.

Nous disposions de la documentation suivante pour le contenu de chaque champ du dataset de Info flora
- "/data/documentation infoflora.xlsx"

## Processus

Dans un premier temps, nous allons nettoyer et trier les données de chaque dataset, pour ne garder que celles qui nous intéressent. Dans un deuxième temps, nous allons comparer les deux datasets afin de déterminer lequel des deux contient le plus de données et si la répartition sur le temps est intéressante.

# Premier dataset : Berce du Caucase

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

df_berce = pd.read_excel('./data/berce_du_caucase.xlsx')

In [3]:
print("Nombre d'enregistrements :", len(df_berce))
print("Nombre de colonnes :",len(df_berce.columns))

pd.set_option('display.max_columns', None) # pour afficher toutes les colonnes
print("Aperçu des 3 premières lignes :")
df_berce.head(10)

Nombre d'enregistrements : 26611
Nombre de colonnes : 50
Aperçu des 3 premières lignes :


gbifID                            datasetKey            occurrenceID  \
0  2881063189  83fdfd3d-3a25-4705-9fbe-3db1d1892b13  INFOFLORA-TRAC-4707071   
1  2882060319  83fdfd3d-3a25-4705-9fbe-3db1d1892b13  INFOFLORA-TRAC-6427214   
2  2881984429  83fdfd3d-3a25-4705-9fbe-3db1d1892b13  INFOFLORA-TRAC-6427213   
3  2881259938  83fdfd3d-3a25-4705-9fbe-3db1d1892b13  INFOFLORA-TRAC-3964916   
4  2881092063  83fdfd3d-3a25-4705-9fbe-3db1d1892b13  INFOFLORA-TRAC-3964917   
5  2880175936  83fdfd3d-3a25-4705-9fbe-3db1d1892b13  INFOFLORA-TRAC-3965118   
6  2881924751  83fdfd3d-3a25-4705-9fbe-3db1d1892b13  INFOFLORA-TRAC-5840993   
7  2881866779  83fdfd3d-3a25-4705-9fbe-3db1d1892b13  INFOFLORA-TRAC-5840992   
8  2879909494  83fdfd3d-3a25-4705-9fbe-3db1d1892b13   INFOFLORA-TRAC-184409   
9  2879040446  83fdfd3d-3a25-4705-9fbe-3db1d1892b13    INFOFLORA-TRAC-12041   

   kingdom        phylum          class    order    family      genus  \
0  Plantae  Tracheophyta  Magnoliopsida  Apiales  Apiaceae  Heracleum   
1  Plantae  Tracheophyta  Magnoliopsida  Apiales  Apiaceae  Heracleum   
2  Plantae  Tracheophyta  Magnoliopsida  Apiales  Apiaceae  Heracleum   
3  Plantae  Tracheophyta  Magnoliopsida  Apiales  Apiaceae  Heracleum   
4  Plantae  Tracheophyta  Magnoliopsida  Apiales  Apiaceae  Heracleum   
5  Plantae  Tracheophyta  Magnoliopsida  Apiales  Apiaceae  Heracleum   
6  Plantae  Tracheophyta  Magnoliopsida  Apiales  Apiaceae  Heracleum   
7  Plantae  Tracheophyta  Magnoliopsida  Apiales  Apiaceae  Heracleum   
8  Plantae  Tracheophyta  Magnoliopsida  Apiales  Apiaceae  Heracleum   
9  Plantae  Tracheophyta  Magnoliopsida  Apiales  Apiaceae  Heracleum   

                    species  infraspecificEpithet taxonRank  \
0  Heracleum mantegazzianum                   NaN   SPECIES   
1  Heracleum mantegazzianum                   NaN   SPECIES   
2  Heracleum mantegazzianum                   NaN   SPECIES   
3  Heracleum mantegazzianum                   NaN   SPECIES   
4  Heracleum mantegazzianum                   NaN   SPECIES   
5  Heracleum mantegazzianum                   NaN   SPECIES   
6  Heracleum mantegazzianum                   NaN   SPECIES   
7  Heracleum mantegazzianum                   NaN   SPECIES   
8  Heracleum mantegazzianum                   NaN   SPECIES   
9  Heracleum mantegazzianum                   NaN   SPECIES   

                              scientificName  \
0  Heracleum mantegazzianum Sommier & Levier   
1  Heracleum mantegazzianum Sommier & Levier   
2  Heracleum mantegazzianum Sommier & Levier   
3  Heracleum mantegazzianum Sommier & Levier   
4  Heracleum mantegazzianum Sommier & Levier   
5  Heracleum mantegazzianum Sommier & Levier   
6  Heracleum mantegazzianum Sommier & Levier   
7  Heracleum mantegazzianum Sommier & Levier   
8  Heracleum mantegazzianum Sommier & Levier   
9  Heracleum mantegazzianum Sommier & Levier   

                      verbatimScientificName verbatimScientificNameAuthorship  \
0  Heracleum mantegazzianum Sommier & Levier                 Sommier & Levier   
1  Heracleum mantegazzianum Sommier & Levier                 Sommier & Levier   
2  Heracleum mantegazzianum Sommier & Levier                 Sommier & Levier   
3  Heracleum mantegazzianum Sommier & Levier                 Sommier & Levier   
4  Heracleum mantegazzianum Sommier & Levier                 Sommier & Levier   
5  Heracleum mantegazzianum Sommier & Levier                 Sommier & Levier   
6  Heracleum mantegazzianum Sommier & Levier                 Sommier & Levier   
7  Heracleum mantegazzianum Sommier & Levier                 Sommier & Levier   
8  Heracleum mantegazzianum Sommier & Levier                 Sommier & Levier   
9  Heracleum mantegazzianum Sommier & Levier                 Sommier & Levier   

  countryCode locality stateProvince occurrenceStatus  individualCount  \
0          CH      NaN            Vs          PRESENT              NaN   
1          CH      NaN            So          PRESENT              NaN

On a donc 26611 lignes et 50 colonnes. Cependant, comme on peut le voir ci-dessus, il y en a beaucoup qui ne sont pas intéressantes pour notre projet. Nous allons donc en retirer une partie pour ne garder que les champs intéressants.

In [4]:
print("Colonnes :\n\n", df_berce.columns)

columns_to_remove = [
    # Colonnes retirées directement 
    'gbifID',
    'datasetKey',
    'occurrenceID',
    'kingdom',
    'phylum',
    'class',
    'order',
    'family',
    'genus',
    'species',
    'scientificName',
    'verbatimScientificName',
    'verbatimScientificNameAuthorship',
    'infraspecificEpithet',
    'taxonRank',
    'locality',
    'stateProvince',
    'publishingOrgKey',
    'taxonKey',
    'speciesKey',
    'institutionCode',
    'collectionCode',
    'catalogNumber',
    'recordNumber',
    'identifiedBy',
    'license',
    'rightsHolder',
    'recordedBy',
    'mediaType',
    'issue',

    # Colonnes retirées après une observation approfondie du fichier
    'occurrenceStatus', # valeur toujours à "PRESENT"
    'individualCount', # champ toujours vide ou 1 et une seule fois 2
    'coordinatePrecision', # champ toujours vide
    'depth', # champ toujours vide
    'depthAccuracy', # champ toujours vide et inutile vu qu'on ne garde pas 'depth'
    'typeStatus', # champ toujours vide
    'lastInterpreted', # pas utile pour notre projet
    'countryCode', # il n'y a que des specimens Suisse dans ce dataset
    'establishmentMeans' # champ toujours vide
]

df_berce_clean = df_berce.drop(columns=columns_to_remove)
df_berce_clean.head(10)

Colonnes :

 Index(['gbifID', 'datasetKey', 'occurrenceID', 'kingdom', 'phylum', 'class',
       'order', 'family', 'genus', 'species', 'infraspecificEpithet',
       'taxonRank', 'scientificName', 'verbatimScientificName',
       'verbatimScientificNameAuthorship', 'countryCode', 'locality',
       'stateProvince', 'occurrenceStatus', 'individualCount',
       'publishingOrgKey', 'decimalLatitude', 'decimalLongitude',
       'coordinateUncertaintyInMeters', 'coordinatePrecision', 'elevation',
       'elevationAccuracy', 'depth', 'depthAccuracy', 'eventDate', 'day',
       'month', 'year', 'taxonKey', 'speciesKey', 'basisOfRecord',
       'institutionCode', 'collectionCode', 'catalogNumber', 'recordNumber',
       'identifiedBy', 'dateIdentified', 'license', 'rightsHolder',
       'recordedBy', 'typeStatus', 'establishmentMeans', 'lastInterpreted',
       'mediaType', 'issue'],
      dtype='object')


decimalLatitude  decimalLongitude  coordinateUncertaintyInMeters  \
0         46.29982           7.91708                           25.0   
1         47.14918           7.50258                            7.0   
2         47.14918           7.50258                            7.0   
3         46.01516           7.73562                         3000.0   
4         47.22541           7.51125                          707.0   
5         47.22541           7.51125                          707.0   
6         46.75684           7.00016                         1100.0   
7         46.88309           7.09097                          707.0   
8         47.45811           9.16405                          750.0   
9         47.14178           8.53529                         2000.0   

   elevation  elevationAccuracy   eventDate   day  month  year  \
0      658.0                1.0  1899-12-30  30.0   12.0  1899   
1      500.0                NaN  1905-06-29  29.0    6.0  1905   
2      500.0                NaN  1905-06-28  28.0    6.0  1905   
3     1700.0                NaN  1923-09-02   2.0    9.0  1923   
4      500.0                NaN  1928-09-20  20.0    9.0  1928   
5      500.0                NaN  1928-07-04   4.0    7.0  1928   
6      723.0                NaN        1930   NaN    NaN  1930   
7      464.0                NaN        1930   NaN    NaN  1930   
8      539.5               56.5        1938   NaN    NaN  1938   
9      542.0                NaN  1948-06-14  14.0    6.0  1948   

        basisOfRecord dateIdentified  
0   HUMAN_OBSERVATION            NaT  
1   HUMAN_OBSERVATION            NaT  
2   HUMAN_OBSERVATION            NaT  
3  PRESERVED_SPECIMEN            NaT  
4  PRESERVED_SPECIMEN            NaT  
5  PRESERVED_SPECIMEN            NaT  
6   HUMAN_OBSERVATION            NaT  
7   HUMAN_OBSERVATION            NaT  
8   HUMAN_OBSERVATION            NaT  
9   HUMAN_OBSERVATION            NaT

In [5]:
# Recherche de valeurs NaN
print("Les colonnes suivantes contiennent des valeurs NaN :")
df_berce_clean.isna().any()

Les colonnes suivantes contiennent des valeurs NaN :


decimalLatitude                  False
decimalLongitude                 False
coordinateUncertaintyInMeters     True
elevation                         True
elevationAccuracy                 True
eventDate                        False
day                               True
month                             True
year                             False
basisOfRecord                    False
dateIdentified                    True
dtype: bool

## Traitement des valeurs NaN
Comme on peut le voir, de nombreuses colonnes contiennent des valeurs NaN. En particulier :
- L'incertitude de la position
- L'altitude (elevation) et la précision
- Le jour et le mois de l'observation
- Le champ "dateIdentified"

## Transformations à effectuer

Certains des champs conservés pourraient donner lieu à d'autres transformations :
### Recherche de valeurs NaN
Il faut vérifier si certains champs ont des valeurs NaN et, le cas échéant, voir si il est possible de définir une règle simple pour savoir si on garde l'enregistrement ou pas.

### Incertitude des coordonnées (**coordinateUncertaintyInMeters**)
On pourrait décider de retirer tous les enregistrements pour lesquels l'incertitude est supérieure à un certain seuil.

### Base d'enregistrement (**basisOfRecord**)
Ce champ indique comment le specimen a pu être identifié. Certains specimen ont la valeur "*PRESERVED_SPECIMEN*", ce qui veut dire qu'ils sont volontairement préservés. Typiquement, il ne faudrait pas prendre ces données si l'on souhaite mesurer l'invasion d'une espèce.

### Dates d'observation
Il y a plusieurs champs correspondant à la date : **eventDate** et le trio : **day**, **month**, **year** ainsi que **dateIdentified**.

C'est problématique car on se sait pas laquelle utiliser. Il faudrait vérifier si ces 3 dates sont toujours équivalentes et, dans le cas contraire, choisir la plus ancienne car elle correspondra forcément à la première observation.

In [6]:
# Pour retirer tous les enregistrements avec une incertitude de position supérieur à un certain seuil

def remove_specimens_with_uncertain_position(df, threshold):
    rows = df['coordinateUncertaintyInMeters'] > threshold
    return df[~rows]


POSITION_UNCERTAINTY_THRESHOLD = 100

count_before_removal = len(df_berce_clean)
df_berce_clean = remove_specimens_with_uncertain_position(df_berce_clean, POSITION_UNCERTAINTY_THRESHOLD)
count_after_removal = len(df_berce_clean)
removed_count = count_before_removal - count_after_removal
print("Nombre d'enregistrements retirés :", removed_count, " (", removed_count / count_before_removal, "% of total rows )")

# Suite à ça, on peut donc également se débarasser de la colonne 'coordinateUncertaintyInMeters'
df_berce_clean.drop(columns=['coordinateUncertaintyInMeters'], inplace=True)


Nombre d'enregistrements retirés : 933  ( 0.035060689188681375 % of total rows )


In [7]:
# On fait la même chose mais pour l'élévation (altitude)

def remove_specimens_with_uncertain_elevation(df, threshold):
    rows = df['elevationAccuracy'] > threshold
    return df[~rows]


MIN_ELEVATION_ACCURACY = 100

count_before_removal = len(df_berce_clean)
df_berce_clean = remove_specimens_with_uncertain_elevation(df_berce_clean, MIN_ELEVATION_ACCURACY)
count_after_removal = len(df_berce_clean)
removed_count = count_before_removal - count_after_removal
print("Nombre d'enregistrements retirés :", removed_count, " (", removed_count / count_before_removal, "% of total rows )")

# Suite à ça, on peut donc également se débarasser de la colonne 'elevationAccuracy'
df_berce_clean.drop(columns=['elevationAccuracy'], inplace=True)

Nombre d'enregistrements retirés : 0  ( 0.0 % of total rows )


In [8]:
# Pour retirer les enregistrements correspondant aux specimens volontairement préservés
def remove_non_invasive_specimens(df):
    rows = df['basisOfRecord'] == "PRESERVED_SPECIMEN"
    return df[~rows]

count_before_removal = len(df_berce_clean)
df_berce_clean = remove_non_invasive_specimens(df_berce_clean)
count_after_removal = len(df_berce_clean)
removed_count = count_before_removal - count_after_removal
print("Nombre d'enregistrements retirés :", removed_count, " (", removed_count / count_before_removal, "% of total rows )")

# On peut donc aussi retirer cette colonne
df_berce_clean.drop(columns=['basisOfRecord'], inplace=True)

Nombre d'enregistrements retirés : 3  ( 0.00011683152893527533 % of total rows )


## Vérification de la cohérence des dates

Le code ci-dessous permet de déterminer si les colonnes eventDate et year,month,day sont incohérentes ou non.

In [9]:
import datetime

# Analyse des champs dates
df_berce_clean[['eventDate', 'day', 'month', 'year', 'dateIdentified']].head(100)

# Première question : Est-ce que eventDate est toujours identique au trio (day, month, year) ?

special_records = []

for idx, record in df_berce_clean[['eventDate', 'day', 'month', 'year']].iterrows():
    eventDate = record['eventDate']
    day = record['day']
    month = record['month']
    year = record['year']

    dt = pd.to_datetime(eventDate)

    if not pd.isna(year) and year != dt.year:
        special_records.append(record)
    elif not pd.isna(month) and month != dt.month:
        special_records.append(record)
    elif not pd.isna(day) and day != dt.day:
        special_records.append(record)

df_special_records = pd.DataFrame(special_records, columns=['eventDate', 'day', 'month', 'year'])
df_special_records.head()

if len(df_special_records) != 0:
    print("Il y a des incohérences entre le champ eventDate et les 3 colonnes year,month,day.")
else:
    print("Pas d'incohérence trouvées au niveau des dates.")

Pas d'incohérence trouvées au niveau des dates.


## Champ dateIdentified

Quant au champ *dateIdentified*, il correspond à une autre information que la date d'observation. Nous n'allons pas le garder.

In [10]:
# On ne garde que le champ 'eventDate' complet
df_berce_clean.drop(columns=['year', 'month', 'day', 'dateIdentified'], inplace=True)

In [11]:
df_berce_clean.head(10)

decimalLatitude  decimalLongitude  elevation   eventDate
0          46.29982           7.91708      658.0  1899-12-30
1          47.14918           7.50258      500.0  1905-06-29
2          47.14918           7.50258      500.0  1905-06-28
14         46.39338           6.99140     1448.0  1970-01-01
15         46.34986           7.16844     1246.5  1970-01-01
19         47.33901           9.18746      900.0        1971
27         47.48006           8.10727      520.0        1980
28         46.90811           9.80881      935.0        1980
33         47.44135           7.88142      546.5  1985-06-19
34         47.45642           8.11227      690.0        1985

## Normalisation des dates

Le champ eventDate n'est pas très pratique car certaines observation n'ont pas la même quantité de données. En effet, parfois nous ne disposons que de l'année alors que dans d'autres cas on dispose d'un timestamp précis à la seconde près.

Nous allons évaluer toutes les dates pour qu'elles aient toutes le même format "année-mois-jour". Si le jour est inconnu, le 1er janvier sera présumé.

In [17]:
df_berce_clean['date'] = pd.to_datetime(df_berce_clean['eventDate'], format='ISO8601')
df_berce_clean.drop(columns=['eventDate'], inplace=True)

df_berce_clean.head(10)

KeyError: 'eventDate'

## Quantités de données par année

todo

## Propriétés du fichier

# Deuxième dataset : Infoflora

Par chance, on a de la documentation pour celui-ci : https://www.infoflora.ch/fr/assets/content/documents/Demande_donnees/legende_fr.pdf

In [38]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

df_flora = pd.read_csv('./data/Infoflora_extraction2.csv')

df_flora.head(5)

C:\Users\lcsch\AppData\Local\Temp\ipykernel_26108\227504881.py:5: DtypeWarning: Columns (0,2,18,21,22,23,24,30) have mixed types. Specify dtype option on import or set low_memory=False.
  df_flora = pd.read_csv('./data/Infoflora_extraction2.csv')


interpretation_note                                taxon management  \
0                 NaN  Ailanthus altissima (Mill.) Swingle        NaN   
1                 NaN  Ailanthus altissima (Mill.) Swingle        NaN   
2       low precision  Ailanthus altissima (Mill.) Swingle        NaN   
3                 NaN  Ailanthus altissima (Mill.) Swingle        NaN   
4       low precision  Ailanthus altissima (Mill.) Swingle        NaN   

   control_type  eradication  doubt_status presence introduction  obs_day  \
0           NaN          NaN             0        +            X      NaN   
1           NaN          NaN             0        +            X      NaN   
2           NaN          NaN             0        +            X      NaN   
3           NaN          NaN             0        +            X      NaN   
4           NaN          NaN             1        +            N      NaN   

   obs_month  obs_year                         observers country canton  \
0        NaN      1969                      Paroz Robert      CH     NE   
1        NaN      1984                      Paroz Robert      CH     NE   
2        NaN      1984                      Paroz Robert      CH     NE   
3        NaN      1989  Duckert-Henriod Marie-Marguerite      CH     NE   
4        NaN      1992                     Clot François      CH     VD   

        municipality      x_swiss      y_swiss  xy_radius  \
0          Neuchâtel  2556500.041  1203499.947      707.0   
1  La Grande Béroche  2548000.041  1191999.947      500.0   
2          Milvignes  2554000.041  1201999.947     1414.0   
3          Neuchâtel  2561000.041  1204899.947      250.0   
4           Lausanne  2538000.041  1151999.947     9999.0   

                                  locality_descript  altitude_min  \
0              Cormondrèche, talus de chemin de fer         560.0   
1                                               NaN         490.0   
2                                               NaN         521.0   
3  à l'angle des rues des Sablons et de Comba-Borel         438.0   
4                                               NaN           NaN   

   altitude_max abundance_cat count_unit abundance cover_code        remarks  \
0           NaN           NaN        NaN       NaN        NaN            NaN   
1           NaN           NaN        NaN       NaN        NaN            NaN   
2           NaN           NaN        NaN       NaN        NaN            NaN   
3         552.0           NaN        NaN       NaN        NaN  Coupé en 1990   
4           NaN           NaN        NaN       NaN        NaN            NaN   

   typo_ch  phenology_code                                  citation  obs_id  \
0      NaN             NaN                                 InfoFlora  184313   
1      NaN             NaN                                 InfoFlora  184312   
2      NaN             NaN                                 InfoFlora  184314   
3      NaN             NaN                                 InfoFlora  184315   
4      NaN             NaN  InfoFlora (Archiv Atlas Welten & Sutter)    1862   

  specimen_type invasive                      taxon_link  sisf_id  taxon_id  \
0           NaN      INV  www.infoflora.ch/flora/1001150     9500   1001150   
1           NaN      INV  www.infoflora.ch/flora/1001150     9500   1001150   
2           NaN      INV  www.infoflora.ch/flora/1001150     9500   1001150   
3           NaN      INV  www.infoflora.ch/flora/1001150     9500   1001150   
4           NaN      INV  www.infoflora.ch/flora/1001150     9500   1001150   

          family     name_de  name_fr  name_it  
0  Simaroubaceae  Götterbaum  Ailante  Ailanto  
1  Simaroubaceae  Götterbaum  Ailante  Ailanto  
2  Simaroubaceae  Götterbaum  Ailante  Ailanto  
3  Simaroubaceae  Götterbaum  Ailante  Ailanto  
4  Simaroubaceae  Götterbaum  Ailante  Ailanto

In [39]:
print("Colonnes :\n\n", df_flora.columns)

columns_to_remove = [
    # Colonnes retirées directement 
    #'taxon', # nom scientifique complet + nom auteur
    'control_type', # methode de lutte 
    'eradication', # état de l'espèce
    'management', # lutte ou non
    'introduction', # comment la plante a été introduite
    'observers', # nom(s) du ou des observateurs
    'country', # toujours égal à CH
    'canton',
    'municipality',
    'locality_descript', # description de l'endroit
    'abundance_cat', # abondance des specimens, souvent vide
    'count_unit', # relatif à l'abondance
    'abundance', # informations supplémentaires par rapport à la taille du specimen
    'cover_code', # souvent vide
    'remarks', # remarques supplémentaires
    'typo_ch', # type d'habitatsss
    'phenology_code', # stade de développement de la plante
    'citation', # citation de la source de données
    'obs_id', # id de l'observation
    'specimen_type', # type d'observation
    'invasive', # redondant, car toutes les espèces de ce dataset sont envahissantes
    'taxon_link', # doc plante
    'sisf_id', # id taxonomique
    'taxon_id', # id taxon
    'family',
    'name_de',
    'name_fr',
    'name_it'
]

df_flora_clean = df_flora.drop(columns=columns_to_remove)
df_flora_clean.head(10)

Colonnes :

 Index(['interpretation_note', 'taxon', 'management', 'control_type',
       'eradication', 'doubt_status', 'presence', 'introduction', 'obs_day',
       'obs_month', 'obs_year', 'observers', 'country', 'canton',
       'municipality', 'x_swiss', 'y_swiss', 'xy_radius', 'locality_descript',
       'altitude_min', 'altitude_max', 'abundance_cat', 'count_unit',
       'abundance', 'cover_code', 'remarks', 'typo_ch', 'phenology_code',
       'citation', 'obs_id', 'specimen_type', 'invasive', 'taxon_link',
       'sisf_id', 'taxon_id', 'family', 'name_de', 'name_fr', 'name_it'],
      dtype='object')


interpretation_note                                taxon  doubt_status  \
0                 NaN  Ailanthus altissima (Mill.) Swingle             0   
1                 NaN  Ailanthus altissima (Mill.) Swingle             0   
2       low precision  Ailanthus altissima (Mill.) Swingle             0   
3                 NaN  Ailanthus altissima (Mill.) Swingle             0   
4       low precision  Ailanthus altissima (Mill.) Swingle             1   
5                 NaN  Ailanthus altissima (Mill.) Swingle             0   
6       low precision  Ailanthus altissima (Mill.) Swingle             0   
7                 NaN  Ailanthus altissima (Mill.) Swingle             1   
8                 NaN  Ailanthus altissima (Mill.) Swingle             1   
9                 NaN  Ailanthus altissima (Mill.) Swingle             1   

  presence  obs_day  obs_month  obs_year      x_swiss      y_swiss  xy_radius  \
0        +      NaN        NaN      1969  2556500.041  1203499.947      707.0   
1        +      NaN        NaN      1984  2548000.041  1191999.947      500.0   
2        +      NaN        NaN      1984  2554000.041  1201999.947     1414.0   
3        +      NaN        NaN      1989  2561000.041  1204899.947      250.0   
4        +      NaN        NaN      1992  2538000.041  1151999.947     9999.0   
5        +      NaN        NaN      1993  2563500.041  1206399.947      250.0   
6        +      NaN        NaN      1994  2570000.041  1210999.947     1414.0   
7        +      NaN        NaN      1996  2538500.001  1153500.001      707.0   
8        +      NaN        NaN      1996  2537500.001  1154500.001      707.0   
9        +      NaN        NaN      1996  2535500.001  1153500.001      707.0   

   altitude_min  altitude_max  
0         560.0           NaN  
1         490.0           NaN  
2         521.0           NaN  
3         438.0         552.0  
4           NaN           NaN  
5         456.0         600.0  
6         431.0           NaN  
7         514.0         663.0  
8         567.0         609.0  
9         396.0         473.0

# Colonnes à utiliser dans la sélection des données

- `interpretation_note` : Donne des informations sur la qualité des données
- `doubt_status` : doute sur l'identification de la plante
- `presence` : indique si la plante est toujours là
- `xy_radius` : précision des coordonnées
- `altitude_min` et `altitude_max` : précision de l'altitude

todo : sélectionner les lignes pertinentes et prendre l'espèce la plus abondante

In [40]:
# 1. Retirer les données douteuses ou invalides
def remove_bad_data(df):
    invalid_tags = [
        'doubtful',
        'not valid',
        'absence',
        'introduced/cultivated',
        #'low precision' # à voir si nécessaire
    ]
    rows = df['interpretation_note'].isin(invalid_tags)
    return df[~rows]

count_before_removal = len(df_flora_clean)
df_flora_clean = remove_bad_data(df_flora_clean)
count_after_removal = len(df_flora_clean)
removed_count = count_before_removal - count_after_removal
print("Enregistrements invalides retirés :", removed_count, " (", removed_count / count_before_removal, "% of total rows )")

# 2. Retirer les specimens dont l'identification est incertaine
def remove_badly_identified_specimens(df):
    invalid_codes = ['2','3']
    rows = df['doubt_status'].isin(invalid_codes)
    return df[~rows]

count_before_removal = len(df_flora_clean)
df_flora_clean = remove_badly_identified_specimens(df_flora_clean)
count_after_removal = len(df_flora_clean)
removed_count = count_before_removal - count_after_removal
print("Specimens mal identifiés retirés :", removed_count, " (", removed_count / count_before_removal, "% of total rows )")

# 3. Retirer les observations qui ne sont plus à jour
def remove_outdated_observations(df):
    outdated_codes = ['-', 'I']
    rows = df['presence'].isin(outdated_codes)
    return df[~rows]

count_before_removal = len(df_flora_clean)
df_flora_clean = remove_outdated_observations(df_flora_clean)
count_after_removal = len(df_flora_clean)
removed_count = count_before_removal - count_after_removal
print("Observations plus à jour retirées :", removed_count, " (", removed_count / count_before_removal, "% of total rows )")

# 4. Retirer les données pas précises
def remove_unprecise_coords(df):
    rows = df['xy_radius'] > POSITION_UNCERTAINTY_THRESHOLD
    return df[~rows]

def remove_unprecise_elevations(df):
    rows = (df['altitude_max'] - df['altitude_min']) > MIN_ELEVATION_ACCURACY
    return df[~rows]

count_before_removal = len(df_flora_clean)
df_flora_clean = remove_unprecise_coords(df_flora_clean)
count_after_removal = len(df_flora_clean)
removed_count = count_before_removal - count_after_removal
print("Observations avec coordonnées XY pas précises :", removed_count, " (", removed_count / count_before_removal, "% of total rows )")

count_before_removal = len(df_flora_clean)
df_flora_clean = remove_unprecise_elevations(df_flora_clean)
count_after_removal = len(df_flora_clean)
removed_count = count_before_removal - count_after_removal
print("Observations avec altitude pas précise :", removed_count, " (", removed_count / count_before_removal, "% of total rows )")

Enregistrements invalides retirés : 2118  ( 0.031068929603497088 % of total rows )
Specimens mal identifiés retirés : 0  ( 0.0 % of total rows )
Observations plus à jour retirées : 0  ( 0.0 % of total rows )
Observations avec coordonnées XY pas précises : 1911  ( 0.028931312733713837 % of total rows )
Observations avec altitude pas précise : 3  ( 4.6771226341554674e-05 % of total rows )


In [41]:
# Retirer les colonnes dont on a plus besoin
df_flora_clean.drop(columns=['interpretation_note'], inplace=True)
df_flora_clean.drop(columns=['doubt_status'], inplace=True)
df_flora_clean.drop(columns=['presence'], inplace=True)
df_flora_clean.drop(columns=['xy_radius'], inplace=True)

In [42]:
df_flora_clean['taxon'].value_counts()
df_flora_clean.head(10)

taxon  obs_day  obs_month  obs_year  \
71   Ailanthus altissima (Mill.) Swingle      NaN        6.0      1998   
72   Ailanthus altissima (Mill.) Swingle     14.0        6.0      2000   
73   Ailanthus altissima (Mill.) Swingle     14.0        6.0      2000   
119  Ailanthus altissima (Mill.) Swingle     20.0        6.0      2002   
120  Ailanthus altissima (Mill.) Swingle     13.0        9.0      2002   
127  Ailanthus altissima (Mill.) Swingle     18.0        6.0      2003   
128  Ailanthus altissima (Mill.) Swingle     11.0       10.0      2003   
130  Ailanthus altissima (Mill.) Swingle      6.0        8.0      2003   
141  Ailanthus altissima (Mill.) Swingle      NaN        NaN      2003   
143  Ailanthus altissima (Mill.) Swingle      NaN        NaN      2003   

         x_swiss      y_swiss  altitude_min  altitude_max  
71   2550967.041  1147511.947         590.0           NaN  
72   2541700.041  1150699.947         372.0         405.0  
73   2541700.041  1150699.947         390.0           NaN  
119  2565500.041  1127749.947         450.0         520.0  
120  2533699.566  1152650.485         390.0         400.0  
127  2563977.041  1207051.947         570.0           NaN  
128  2507500.041  1137849.947         405.0           NaN  
130  2554000.041  1146299.947         385.0           NaN  
141  2554500.041  1145799.947         392.0         404.0  
143  2558000.041  1143699.947         391.0         400.0

## Normalisation des dates

Comme pour l'autre fichier sur les Berces, il convient de mettre toutes les dates au même format.

In [46]:
df_flora_clean.rename(columns={'obs_day':'day','obs_month':'month','obs_year':'year'}, inplace=True)
df_flora_clean.fillna(value={'day':1, 'month':1}, inplace=True)

df_flora_clean['date'] = pd.to_datetime(df_flora_clean[['year','month','day']])
df_flora_clean.drop(columns=['year','month','day'], inplace=True)

df_flora_clean.head(10)

taxon      x_swiss      y_swiss  \
71   Ailanthus altissima (Mill.) Swingle  2550967.041  1147511.947   
72   Ailanthus altissima (Mill.) Swingle  2541700.041  1150699.947   
73   Ailanthus altissima (Mill.) Swingle  2541700.041  1150699.947   
119  Ailanthus altissima (Mill.) Swingle  2565500.041  1127749.947   
120  Ailanthus altissima (Mill.) Swingle  2533699.566  1152650.485   
127  Ailanthus altissima (Mill.) Swingle  2563977.041  1207051.947   
128  Ailanthus altissima (Mill.) Swingle  2507500.041  1137849.947   
130  Ailanthus altissima (Mill.) Swingle  2554000.041  1146299.947   
141  Ailanthus altissima (Mill.) Swingle  2554500.041  1145799.947   
143  Ailanthus altissima (Mill.) Swingle  2558000.041  1143699.947   

     altitude_min  altitude_max       date  
71          590.0           NaN 1998-06-01  
72          372.0         405.0 2000-06-14  
73          390.0           NaN 2000-06-14  
119         450.0         520.0 2002-06-20  
120         390.0         400.0 2002-09-13  
127         570.0           NaN 2003-06-18  
128         405.0           NaN 2003-10-11  
130         385.0           NaN 2003-08-06  
141         392.0         404.0 2003-01-01  
143         391.0         400.0 2003-01-01

## Propriétés du fichier

## Comparaison des deux fichiers

todo : quantités d'observations par année, nombre de plantes différentes, etc...

# Conclusion